📖 WIP: Label Sync Between MachineDeployment and underlying Kubernetes Nodes (#6255)
kubernetes-sigs/cluster-api: docs/proposals/20220210-md-node-label-sync.md

- Labels are only applicable at creation time. Adding/Removing a label would require a MachineDeployment rollout. This is undesirable especially in baremetal environments where provisioning can be time consuming. 

The documented approach for specifying restricted labels in CAPI is to utilize kubectl. This introduces the potential for human error and also goes against the general declarative model adopted by CAPI.

Users could also implement their own label synchronizer in their tooling, but this may not be practical for most. 

### Goals

- Support synchronizing labels that fall under the following restricted prefixes: `node-role.kubernetes.io/*` and `node-restriction.kubernetes.io/*`
- Support synchronizing labels that fall under a CAPI specific prefix: `node.cluster.k8s.io/*`

### Non-Goals/Future Work

- Label sync for control-plane nodes. 
- Support for arbitrary/user-specified label prefixes. 

## Proposal

CAPI will define its own label prefix which falls within the [restricted namespace *.k8s.io/](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/279-limit-node-access):

```
node.cluster.k8s.io/*
```

> Labels placed on the Machine that match the above prefixe will be authoritative -- Cluster API will act as the ultimate authority for anything under the prefix. Labels with the above prefix should be set only within Cluster API CRDs.

Reviewer: typo "prefixe"

#### Story 2

As a cluster admin/user, for the purpose of workload placement, I would like a declarative and secure means by which I can add/remove labels on groups of nodes. For example, I want to run my outward facing nginx pods on a specific set of nodes (e.g. md-secure-zone) to reduce exposure to the company's publicly trusted server certificates.

### Implementation Details/Notes/Constraints

#### Label propogation from MachineDeployment to Machine

Option #1: CAPI metadata Propagation

CAPI supports [metadata propagation](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/metadata-propagation.html?highlight=propagation#metadata-propagation) across core API resources. For the MachineDeployment types, template labels propagate to MachineSets top-level and MachineSets template metadata:
```
MachineDeployment.spec.template.metadata.labels => MachineSet.labels, MachineSet.spec.template.metadata.labels
```
Similarly, on MachineSet, template labels propagate to Machines:
```
MachineSet.spec.template.metadata.labels => Machine.labels
```

> As of this writing, changing the `MachineDeployment.spec.template.metadata.labels` will trigger a rollout. This is undesirable. 

Reviewer: I'd suggest to keep this a the first option and update it to propose adding support in the MD to recognise these labels and exclude them from the comparison for upgrades similar to what we do for MachineDeploymentUniqueLabel and move the other options under the alternative sections.

> As of this writing, changing the `MachineDeployment.spec.template.metadata.labels` will trigger a rollout. This is undesirable. 

Author: @enxebre That's a good idea!
Looking at the current code, it would mean skipping the comparison of these specific prefixes in `EqualMachineTemplate()`. 

Any other gotchas here?

> As of this writing, changing the `MachineDeployment.spec.template.metadata.labels` will trigger a rollout. This is undesirable. 

Reviewer 2: The main gotcha with this stuff is that we have to make it backwards compatible (i.e. a rollout of a CAPI version with the new hash calculation should not automatically trigger rollouts).

But not sure if you have to mention that in the proposal.

```

Option #2: New field Spec.NodeLifecycle.Labels

Instead of using the top-level metadata, a new field would be introduced:
```
MachineDeployment.Spec.NodeLifecycle.Labels => MachineSet.Spec.NodeLifecycle.Labels => Machine.Spec.NodeLifecycle.Labels
```

The benefit being that it provides a clear indication to the user that these labels will be synced to the Kubernetes Node(s). This will, however, require api changes. 

#### Synchronization of CAPI Labels

Labels that fall under the CAPI specific prefix `node.cluster.k8s.io/*` will need to be kept in sync between the Machine and Node objects. Synchronization must handle these two scenarios: 
(A) Label is added/removed on a Machine and must be added/removed on the corresponding Node.
(B) Label is added/removed on the Node and must be removed/re-added on the Node to bring it in sync with the labels on Machine.

#### Machine Controller Changes

On each Machine reconcile event, the machine controller will fetch the corresponding Node (i.e. Machine.Status.NodeRef). If the Machine has labels specified, then the controller will ensure those labels are applied to the Node. 

> To watch for changes on the Node object, the Tracker pattern will be used. ...

Author: Done. Thanks.

---

title: Label Sync Between MachineDeployment and underlying Kubernetes Nodes
authors:
  - "@arvinderpal"
  - "@enxebre"
reviewers:
  - "@fabriziopandini"
  - "@neolit123"
  - "@sbueringer"
  - "@vincepri"
  - "@randomvariable"
creation-date: 2022-02-10
last-updated: 2022-02-26
status: implementable
replaces:
superseded-by:

---

# Label Sync Between MachineDeployment and underlying Kubernetes Nodes

## Glossary

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

## Summary

Labels on worker nodes are commonly used for node role assignment and workload placement. This document discusses a means by which labels placed on MachineDeployments can be kept in sync with labels on the underlying Kubernetes Nodes. 

## Motivation

Kubelet supports self-labeling of nodes via the `--node-labels` flag. CAPI users can specify these labels via `kubeletExtraArgs.node-labels` in the KubeadmConfigTemplate. There are a few shortcomings: 

> - Security concerns have restricted the allowed set of labels; In particular, kubelet's ability to self label in the `*.kubernetes.io/` and `*.k8s.io/` namespaces is mostly prohibited. Using prefixes outside these namespaces (i.e. unrestricted prefixes) is discouraged.  

Reviewer 2: Sorry my comment is confusing.

I think from a Kubernetes perspective it's fine to use node-role and node-restriction, but apart from that only Kubernetes core components should use *.k8s.io and *.kubernetes.io. As we are not a Kubernetes core component we have to user our own namespace anyway.

> - Security concerns have restricted the allowed set of labels; In particular, kubelet's ability to self label in the `*.kubernetes.io/` and `*.k8s.io/` namespaces is mostly prohibited. Using prefixes outside these namespaces (i.e. unrestricted prefixes) is discouraged.  

Reviewer 2: So in general that point you made is fine. I just think "Using prefixes outside these namespaces (i.e. unrestricted prefixes) is discouraged." is counter-intuitive as my interpretation of upstream policies is that we **have** to use our own namespace (at least for our own labels)

- Labels are only applicable at creation time. Adding/Removing a label would require a MachineDeployment rollout. This is undesirable especially in baremetal environments where provisioning can be time consuming. 

The documented approach for specifying restricted labels in CAPI is to utilize kubectl. This introduces the potential for human error and also goes against the general declarative model adopted by CAPI.

Users could also implement their own label synchronizer in their tooling, but this may not be practical for most. 

### Goals

- Support synchronizing labels that fall under the following restricted prefixes: `node-role.kubernetes.io/*` and `node-restriction.kubernetes.io/*`
- Support synchronizing labels that fall under a CAPI specific prefix: `node.cluster.k8s.io/*`

> ### Non-Goals/Future Work

Reviewer 2: Okay pending discussion above Control Plane might move to goals but MachinePool machines should stay in future work (as MachinePools today don't have Machines pending the open proposal)

> - Label sync for control-plane nodes. 

Reviewer 3: ```suggestion
- Label sync for control-plane nodes and for MachinePools nodes. 
```

- Support for arbitrary/user-specified label prefixes. 

## Proposal

CAPI will define its own label prefix which falls within the [restricted namespace *.k8s.io/](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/279-limit-node-access):

```

> node.cluster.k8s.io/*

Reviewer 3: Most of the CAPI labels are under `x-k8s.io`. Should we use this here as well?

Reviewer 2: I think we're mixing up a few things. There are security concerns about the labels that kubelet sets. Specifically kubelet itself is only allowed to set a few labels because otherwise malicious kubelets could self-label and thus do bad things.

But we're discussion which labels we are setting as CAPI and I think in this case the following applies:
> The kubernetes.io/ and k8s.io/ prefixes are [reserved](https://kubernetes.io/docs/reference/labels-annotations-taints/) for Kubernetes core components.
> https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

As we are not a Kubernetes core component we have to use our own namespace (like we did for all our other labels)

P.S. The comment order is interesting :) (yours 4 hours ago is above mine with 11 hours ago)

Author: I'll change the above to:
```
node.cluster.x-k8s.io/*
```

Reviewer 2: Sounds good!

```

Labels placed on the Machine that match the above prefix will be authoritative -- Cluster API will act as the ultimate authority for anything under the prefix. Labels with the above prefix should be set only within Cluster API CRDs.

> In addition to the above, CAPI will support one-time application of the following labels during Machine creation:

Author: Yes, that's a good idea. I will add that to the doc.

> In addition to the above, CAPI will support one-time application of the following labels during Machine creation:

Reviewer 2: Feel free to resolve if it has been added

```
MachineDeployment.spec.template.metadata.labels => MachineSet.labels, MachineSet.spec.template.metadata.labels
```
Similarly, on MachineSet, template labels propagate to Machines:
```
MachineSet.spec.template.metadata.labels => Machine.labels
```

As of this writing, changing the `MachineDeployment.spec.template.metadata.labels` will trigger a rollout. This is undesirable. CAPI will be updated to instead ignore updates of labels that fall within the CAPI prefix. There is precedence for this with the handling of `MachineDeploymentUniqueLabel`. 

#### Synchronization of CAPI Labels

Labels that fall under the CAPI specific prefix `node.cluster.k8s.io/*` will need to be kept in sync between the Machine and Node objects. Synchronization must handle these two scenarios: 
(A) Label is added/removed on a Machine and must be added/removed on the corresponding Node.
(B) Label is added/removed on the Node and must be removed/re-added on the Node to bring it in sync with the labels on Machine.

> #### Machine Controller Changes

Reviewer 3: Given that this proposal is implemented at Machine controller level, I assume that the gap for extending this to control plane machines is minimal. Can we reconsider including this in the proposal scope?

Reviewer 2: +1 I think it's easier to implement this feature for all Machines (CP & MachinePool) vs. implementing something that it only works for MachineSet/MachineDeployment managed machines.

Author: More than happy to expand this to CP, but I wonder if we can first get this implemented for workers and work out any kinks along the way. The proposal can then be easily expanded to include CP.

Reviewer 2: I think it's the other way around. If you want to only sync labels from "MachineDeployment Machines" you have to implement handling in the machine controller to figure out if the current Machine is managed by a MD/MS, KCP, MachinePool, some third-party tool (@enxebre :)).

I think it's easier to avoid adding something like that to the machine controller. IIRC the only MD specific change is that we have to avoid triggering rollouts. I think it might be easier to figure out how we make this happen for all "owners" vs how we make it happen that the feature only works for "MD Machines".

Reviewer 2: I think let's for now figure this out how exactly we do this for MD's (https://github.com/kubernetes-sigs/cluster-api/pull/6255#discussion_r831812895) and then we can give a hint how to phrase this for KCP and MachinePools.

I think it comes down to:
1. Implement label sync in the MachinePool controller
2. Implement in-place label propagation MD=>MS=>Machine
3. Implement in-place label propagation KCP=>Machine
4. Implement in-place label propagation MachinePool=>Machine

We don't have to implement 3+4 right away. But I would prefer implementing them as follow-ups to the first implementation vs. having to implement something to block label sync for KCP/MachinePool machines after 1.

Reviewer 2: > Did you mean Machine controller in 1?

Yes, sorry my bad. edited

Reviewer 2: @Arvinderpal what do you think?

Reviewer 4: We'll probably need the first 3 implemented right away to make sure there is basic functionality for this feature, otherwise it doesn't seem most folks would be able to use it?

Reviewer 2: > We'll probably need the first 3 implemented right away to make sure there is basic functionality for this feature, otherwise it doesn't seem most folks would be able to use it?

Depending on complexity it might be a good idea to have 3 separate PRs to implement 1,2,3. The feature becomes really useful after 2,3. But after 1 it's already possible to propagate labels to nodes, just not to update them without a rolling update.

My main point is, I would really like to avoid to implement something in the Machine controller which enables/disables the label sync based on if the Machine is owned by a MachineSet/MachineDeployment or KCP.

Author: Yes, that sounds good to me. I'll update the proposal to incorporate KCP.

Reviewer 3: Given the general agreement on this point, IMO updating the proposal for KCP support is a prerequisite before starting lazy consensus

Author: I have added the following to the goals:
```
- Support label synchronization for MachineDeployments as well as KubeadmControlPlane (KCP).
```

As noted above by @sbueringer we can start with MDs to iron out the details and incorporate the same functionality into KCP in follow up PRs.

On each Machine reconcile event, the machine controller will fetch the corresponding Node (i.e. Machine.Status.NodeRef). If the Machine has labels specified, then the controller will ensure those labels are applied to the Node. 

Similarly, today the machine controller watches Nodes within the workload cluster, where changes to the Node trigger a reconcile of its corresponding Machine object. This should allow us to capture changes made directly to the labels on the Node (e.g. user accidently deletes a label that falls under CAPI's managed prefix).

#### One-time Apply of Standard Kubernetes Labels

> The machine controller will apply the standard kubernetes labels, if specified, to the Node immediately after the Node enters the Ready state (ProviderID is set...).

Reviewer 3: We should make sure this is re-entrant, but we can figure out this during implementation

Reviewer 2: What about adjusting CABPK to pick up the kubernetes.io labels and set them via "--node-labels". This would solve the race condition below (at lesat for those labels) and the "one-time" apply for the kubernetes.io labels would happen when the node joins.

We can recommend similar behavior for other bootstrap providers.

I think it would also simplify the handling of this feature for users (see discussion below around taints)

Author: @sbueringer As documented earlier in the doc, the restricted labels cannot be set via the --node-labels flag.

Reviewer 2: My bad. I thought they are in the allow-list, but they are intentionally not.

#### Delay between Node Create and Label Sync 

> The Node object is first created when kubeadm joins a node to the workload cluster (i.e. kubelet is up and running). There may be a delay (potentially several seconds) before the LSC kicks in to apply the labels on the Node.

Reviewer 3: What is the LSC? (might be necessary to define it or simply use Machine Controller without introducing a new abbreviation)

Kubernetes supports both equality and inequality requirements in label selection. In an equality based selection, the user wants to place a workload on node(s) matching a specific label (e.g. Node.Labels contains `my.prefix/foo=bar`). The delay in applying the label on the node, may cause a subsequent delay in the placement of the workload, but this is likely acceptable.

> In an inequality based selection, the user wants to place a workload on node(s) that do not contain a specific label (e.g. Node.Labels not contain `my.prefix/foo=bar`). The case is potentially problematic because it relies on the absence of a label and this can occur if the pod scheduler runs during the delay interval.

Reviewer 2: I wonder if for cases like this taints are the more appropriate solution.

Reviewer 2: Let's resolve, not super important

> One way to address this is to use kubelet's `--register-with-taints` flag. We could use that to taint a newly minted Node with say `cluster.x-k8s.io=label-sync-pending:NoSchedule`. Assuming workloads don't have this specific toleration, then nothing should be scheduled. When the LSC synchronizes labels, it removes this specific taint.

Reviewer 3: I'm not sure if we are discussing options or if we are documenting somethings that will be implemented by this proposal

KubeadmConfigTemplate provides the means to set taints on nodes (see  JoinConfiguration.NodeRegistration.Taints). We have some options as to how best to “insert” this specific taint: 
(1) Leave it to the user to specify this exact taint in their KubeadmConfigTemplate manifests
(2) CABPK inserts it by default in all join configurations. 

> Option (1) is the simplest and could be documented as so:

Reviewer 3: Am I wrong or this option does not consider removing the taint when the labels are synced

Reviewer 2: I wonder if we should consider just automatically adding the taint. Will just make the UX easier when folks are using CABPK. But yeah if it's an edge case that most users don't care about we can only document it (and depending on demand potentially add it automatically later)

```
If you utilize inequality based selection for workload placement, to prevent unintended scheduling of pods during the initial node startup phase, it is recommend that you specify the following taint in your KubeadmConfigTemplate:
`cluster.x-k8s.io=label-sync-pending:NoSchedule`
```

## Alternatives

### Label propogation from MachineDeployment to Machine

Option #1: Introduce a new field Spec.NodeLifecycle.Labels

Instead of using the top-level metadata, a new field would be introduced:

```

> MachineDeployment.Spec.NodeLifecycle.Labels => MachineSet.Spec.NodeLifecycle.Labels => Machine.Spec.NodeLifecycle.Labels

Reviewer 3: I personally prefer this option given that it introduces a clear and explicit boundary between machine labels and node labels.
However, if we go down this path we should consider also impact on ClusterClass and managed topologies

Reviewer 2: I'm fine with both options. But I think I would use `MachineDeployment.spec.template.node.metadata.labels` instead.

But let's get consensus on this before refactoring the proposal back and forth. I think the last time our tendency (@vincepri) was to re-use the existing labels map.

Author: Yes, during one of our office hours, the desire was to reuse the existing label propagation.

Reviewer: To me `Machine.Spec.NodeLifecycle.Labels` which would also eventually contain drainTimeout and NodeDeletionTimeout is the most intuitive.

Reviewer 2: nit: I think my point is mainly I don't see the point in the "Lifecycle" suffix if labels are part of the node lifecycle, what isn't? :)

Author: I guess we need to reach some agreement on this. Is @vincepri going to make the final call?
Whether we use MD.Medata.Label or MD.spec.template.node.metadata.labels, we'll need to support in-place propagation similar to https://github.com/kubernetes-sigs/cluster-api/issues/5880 , since today both scenarios would trigger a rollout, right?

Reviewer 2: > we'll need to support in-place propagation similar to https://github.com/kubernetes-sigs/cluster-api/issues/5880 , since today both scenarios would trigger a rollout, right?

Yes in both cases we need in-place propagation, but only with the new field we're able to avoid a behavioral change (see https://github.com/kubernetes-sigs/cluster-api/pull/6255/files#r881737137).

I'm now leaning towards having a separate field. I personally would prefer `MachineDeployment.spec.template.node.labels` over something with `lifecycle` as `lifecycle` sounds very very generic to me so I'm not sure if it's beneficial to have it in the field name.

I think @vincepri's opinion would be good to hear

Reviewer 4: If we're planning to expand the `node` field in the future, `node.labels` works for me. I'm also fine re-using the usual labels, and only sync the ones with the pre-defined prefix

Reviewer 2: As written above, my tendency would be to use a separate field to make it clear. But I don't want to block if the consensus is to reuse the existing labels field and only sync some of them (the one defined in this proposal) to nodes.

Reviewer 2: Resolving given we had consensus on reusing the existing labels field

---

title: Label Sync Between MachineDeployment and underlying Kubernetes Nodes
authors:
  - "@arvinderpal"
  - "@enxebre"
reviewers:
  - "@fabriziopandini"
  - "@neolit123"
  - "@sbueringer"
  - "@vincepri"
  - "@randomvariable"
creation-date: 2022-02-10
last-updated: 2022-02-26
status: implementable
replaces:
superseded-by:

---

# Label Sync Between MachineDeployment and underlying Kubernetes Nodes

## Glossary

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

## Summary

Labels on worker nodes are commonly used for node role assignment and workload placement. This document discusses a means by which labels placed on MachineDeployments can be kept in sync with labels on the underlying Kubernetes Nodes. 

## Motivation

> Kubelet supports self-labeling of nodes via the `--node-labels` flag. CAPI users can specify these labels via `kubeletExtraArgs.node-labels` in the KubeadmConfigTemplate. There are a few shortcomings: 

Reviewer: >Security concerns have restricted the allowed...

can you add a link to where this is stated?

Author: Will do!

#### Story 2

As a cluster admin/user, for the purpose of workload placement, I would like a declarative and secure means by which I can add/remove labels on groups of nodes. For example, I want to run my outward facing nginx pods on a specific set of nodes (e.g. md-secure-zone) to reduce exposure to the company's publicly trusted server certificates.

### Implementation Details/Notes/Constraints

#### Label propogation from MachineDeployment to Machine

Utilize CAPI metadata Propagation. CAPI supports [metadata propagation](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/metadata-propagation.html?highlight=propagation#metadata-propagation) across core API resources. For the MachineDeployment types, template labels propagate to MachineSets top-level and MachineSets template metadata:
```
MachineDeployment.spec.template.metadata.labels => MachineSet.labels, MachineSet.spec.template.metadata.labels
```
Similarly, on MachineSet, template labels propagate to Machines:
```
MachineSet.spec.template.metadata.labels => Machine.labels
```

> As of this writing, changing the `MachineDeployment.spec.template.metadata.labels` will trigger a rollout. This is undesirable. CAPI will be updated to instead ignore updates of labels that fall within the CAPI prefix. There is precedence for this with the handling of `MachineDeploymentUniqueLabel`. 

Reviewer 2: Let's consider the following.
* A MD is initially created with some "node labels"
* Labels are propagated MD => MS => Machine
* Machine controller will propagate the "node labels" from the Machine to the node in the workload cluster

Now we modify one of those node labels on the MachineDeployment. With our changes to the MD controller this won't trigger a rollout of MachineSets / Machines (which is intended) but it also won't propagate the changed node labels to the workload nodes as the MachineSet and the Machine is not rolled out and thus the machine controller doesn't know that it has to sync the changed node labels.

I assume this is not intended. Do we need some kind of "inline patching" in the MD and MS controllers? I.e. the MD controller patches the existing MachineSets and the MachineSet controller patches the Machines if node label change? (instead of rolling them out like for all other changes)

Reviewer: Yeh I think this need to be rephrased a bit to clarify we'll introduce in-place reconciliation from the MD->MS-Machines for these labels. Implementation wise this need to be solved generically adding first class support for [in-place propagation which also affect other scenarios](https://github.com/kubernetes-sigs/cluster-api/issues/5880).

Reviewer 2: One thing that we have to keep in mind is that there might be user expectations today that changing labels/annotations triggers rollouts.

I think it's sometimes used as a workaround because rolloutAfter isn't available yet on MDs. We even mention it in our documentation: https://cluster-api.sigs.k8s.io/tasks/upgrading-clusters.html?highlight=rolloutAfter#how-to-schedule-a-machine-rollout

Resolution could be to get the MachineDeployment rolloutAfter in place before merging inline patching of labels/annotations.

Author: Is there a ticket for  MachineDeployment rolloutAfter? It will be a blocker for the inplace propagation of labels and in general seems like a good feature to have.

Reviewer: I'm not sure I follow how is rolloutAfter related? I'd expect us to do this in a way we preserve existing behaviour.

>I assume we would do it for all labels not only the ones for nodes.

Ah I see because of this assumption. I'm not 100% sure on this assumption, may be this is another arg to have a specific field in the API for Node labels.

Reviewer 2: Yup, sounds like a good reason to have a separate field, I think otherwise there is no way to preserve existing behavior.

#### Story 2

As a cluster admin/user, for the purpose of workload placement, I would like a declarative and secure means by which I can add/remove labels on groups of nodes. For example, I want to run my outward facing nginx pods on a specific set of nodes (e.g. md-secure-zone) to reduce exposure to the company's publicly trusted server certificates.

### Implementation Details/Notes/Constraints

#### Label propogation from MachineDeployment to Machine

Utilize CAPI metadata Propagation. CAPI supports [metadata propagation](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/metadata-propagation.html?highlight=propagation#metadata-propagation) across core API resources. For the MachineDeployment types, template labels propagate to MachineSets top-level and MachineSets template metadata:
```
MachineDeployment.spec.template.metadata.labels => MachineSet.labels, MachineSet.spec.template.metadata.labels
```
Similarly, on MachineSet, template labels propagate to Machines:
```
MachineSet.spec.template.metadata.labels => Machine.labels
```

> As of this writing, changing the `MachineDeployment.spec.template.metadata.labels` will trigger a rollout. This is undesirable. CAPI will be updated to instead ignore updates of labels that fall within the CAPI prefix. There is precedence for this with the handling of `MachineDeploymentUniqueLabel`. 

Reviewer 2: ```suggestion
As of this writing, changing the `MachineDeployment.spec.template.metadata.labels` will trigger a rollout. This is undesirable. CAPI will be updated to instead propagate updates of labels in-place without a rollout. Similar changes will be made to KCP.
```

I would suggest to propagate all labels inline. Otherwise we will end up in a weird state where changes to some labels trigger a MD/MS rollout while others are in-place updated (should be simpler to implement as well).

Reviewer 2: Yeah which is why I favored a separate field ... but yeah that decision has been made now :)

Just to confirm, we're fine with breaking behavior changes of our API fields as long as we do them in new minor releases?

Reviewer 4: > Just to confirm, we're fine with breaking behavior changes of our API fields as long as we do them in new minor releases? (my impression was that this requires a new apiVersion)

The API fields wouldn't need to change, but the propagation behavior of those fields would change, which is something that needs to be documented.

> P.S. re: MD changes. The real fun part is probably to stop considering labels for the hash we're using while not triggering a MD rollout when the new CAPI version with this behavior is rolled out

The same problem applies if we add a new field to the Machine template, we'd need a way to make sure that the generated hash is backward compatible. If that's easier though, let's just add a new field?

Reviewer 2: > The API fields wouldn't need to change, but the propagation behavior of those fields would change, which is something that needs to be documented.

Okay, good to know :)

> The same problem applies if we add a new field to the Machine template, we'd need a way to make sure that the generated hash is backward compatible. If that's easier though, let's just add a new field?

I think if it's a separate field we just never have to include it in the hash and we don't have to figure out how to remove something from the hash calculation without rollouts. I think?

Reviewer 2: I did a bit of research:
1. go-spew seems to be basically unmaintained since 4 years (https://github.com/davecgh/go-spew)
2. only a fork of it allows to exclude fields from the hash calculation https://github.com/picmonkey/go-spew/pull/1

Independent of that, hash calculation works something like that:
1. Use spew to write a representation of an object (in our case a MachineTemplateSpec) into a hasher
2. spew writes the object like this: (note the NodeLabels I introduced which are there even though the field is not set)
> &v1beta1.MachineTemplateSpec{ObjectMeta:v1beta1.ObjectMeta{Labels:map[string]string(nil), Annotations:map[string]string(nil)}, Spec:v1beta1.MachineSpec{ClusterName:"abc", NodeLabels:map[string]string(nil), Bootstrap:v1beta1.Bootstrap{ConfigRef:(*v1.ObjectReference)(nil), DataSecretName:(*string)(nil)}, InfrastructureRef:v1.ObjectReference{Kind:"", Namespace:"", Name:"", UID:"", APIVersion:"", ResourceVersion:"", FieldPath:""}, Version:(*string)(nil), ProviderID:(*string)(nil), FailureDomain:(*string)(nil), NodeDrainTimeout:(*v1.Duration)(nil), NodeDeletionTimeout:(*v1.Duration)(nil)}}
3. hasher calculates the hash

Now let's look at our options:

### 1. Reuse the existing labels field but stop rolling out for label changes
(Assuming we don't want to accept a rollout of all MDs after a CAPI upgrade)

Today the hash is calculated including all labels. We would have to find a way to calculate the hash:
* that ignores any label changes
* hash shouldn't change to what we have today
* still rolls out when anything else changes
I might be missing something, but that sounds close to impossible to me.

### 2. Introduce a new field and just never include it in the hash calculation

Based on what I wrote above how the hash calculation works there is at least the following way to do it:
1. Let spew write the object into a buffer
2. drop the new field from the string in the buffer
3. use the resulting string for hash calculation

There are probably alternatives, the best one would be to find a compatible library which just allows us to mark the new field so that it isn't included in the hash calculation (e.g. like the fork of spew)

Opinions? (@enxebre @vincepri)

P.S. sorry don't want to hold the proposal further back, but I'm now absolutely not sure if the current approach is implementable.

Reviewer 4: Between the two solutions proposed above, the second one seems the most sane from a user perspective. If we also think that in the fullness of time we'd want to expand the in place node fields, this might be a good start.

That said, the biggest issue I see with the above is that we're relying on a hashing mechanism that's flaky and fragile. Have we considered breaking the hashing first before introducing new changes that might affect future rollouts and extensibility points?

Reviewer 2: > That said, the biggest issue I see with the above is that we're relying on a hashing mechanism that's flaky and fragile. Have we considered breaking the hashing first before introducing new changes that might affect future rollouts and extensibility points?

Good point. I don't think we went that far regarding thinking about how to implement this proposal. But I think we should try to get to a consensus relatively soon, we're going back and forth on the field discussion since a while.

Maybe the following is a path forward?

1. Pick the option that labels will be in a separate node.labels field
2. Implement the sync from Machine to Nodes
3. Figure out how to in-place upgrade node.labels in MD
4. Figure out how to in-place upgrade node.labels in KCP

I think the feature already would have a big benefit after 2., even without in-place upgrades (as it at least allows to propagate labels to nodes at all, which isn't possible right now).

Then we can continue with 3. and doing POC's and figuring out what the ideal way is to implement in-place upgrades.

Reviewer 2: Damn :/ Forgot that adding the field will already lead to a rollout on CAPI upgrade if we don't change the hashing.
So we definitely have to figure out how to do it for 2. already.

But If the general consensus is that we prefer the new field anyway (and we have the hash problem in both cases) we can at least move the proposal forward and then have to experiment on the implementation how to actually do it.

Reviewer 2: > Should we also consider replacing the [Deepcopy](https://github.com/kubernetes-sigs/cluster-api/blob/bd48f0275d0e3ddae7bcf99d200aa3e940c39ee2/internal/controllers/machinedeployment/machinedeployment_sync.go#L144) with something that includes specific fields (i.e. excludes our new field).

I think we as long as we are using a changed struct the hash will change. I think we either have to:
1. "freeze" the struct by creating another one which produces the same output
2. hack the way we write the struct into the hasher (e.g. by just dropping parts of the string)
3. Replace Spew with something which writes a compatible output but allows ignoring fields
4. Replace the entire hash mechanism through something else

There might be other alternatives as well. But for me personally it would be fine to experiment during implementation. I think by making it a separate field we have a few advantages and one of them is that that hash issue is probably simpler to solve and that it's not a behavioral change of the label field anymore.

> So, when people upgrade, there will always be a rollout? Is this behavior considered okay for a breaking change?

We're trying to avoid that at all costs (we even have a e2e test which validates it)

#### Synchronization of CAPI Labels

> Labels that fall under the CAPI specific prefix `node.cluster.k8s.io/*` will need to be kept in sync between the Machine and Node objects. Synchronization must handle these two scenarios: 
> (A) Label is added/removed on a Machine and must be added/removed on the corresponding Node.
> (B) Label is added/removed on the Node and must be removed/re-added on the Node to bring it in sync with the labels on Machine.

Reviewer 2: ```suggestion
Labels that fall under the CAPI specific prefix `node.cluster.k8s.io/*` will need to be kept in sync between the Machine and Node objects. Synchronization must handle these three scenarios: 
(A) Label is added/removed on a Machine and must be added/removed on the corresponding Node.
(B) Label is added/removed on the Node and must be removed/re-added on the Node to bring it in sync with the labels on Machine.
(C) Label value is changed on the Machine or Node and must be brought in sync with the label value on Machine.
```
I think we have 3 scenarios (or something similar)

#### Machine Controller Changes

On each Machine reconcile event, the machine controller will fetch the corresponding Node (i.e. Machine.Status.NodeRef). If the Machine has labels specified, then the controller will ensure those labels are applied to the Node. 

Similarly, today the machine controller watches Nodes within the workload cluster, where changes to the Node trigger a reconcile of its corresponding Machine object. This should allow us to capture changes made directly to the labels on the Node (e.g. user accidently deletes a label that falls under CAPI's managed prefix).

#### One-time Apply of Standard Kubernetes Labels

> The machine controller will apply the standard kubernetes labels, if specified, to the Node immediately after the Node enters the Ready state (ProviderID is set). We will enforce this one-time application by making the labels immutable via a validating webhook. 

Reviewer 2: @fabriziopandini This could be an interesting edge case if we choose to in-place update all labels from MD=>MS=>M

The plan was to only apply those labels one time on a node and then don't change them later. But if a label change doesn't trigger an entire rollout I'm not sure how we can rollout those labels from a MD to nodes

Reviewer 2: Discussed it. We can solve this either with rolloutAfter or document the limitation

In an inequality based selection, the user wants to place a workload on node(s) that do not contain a specific label (e.g. Node.Labels not contain `my.prefix/foo=bar`). The case is potentially problematic because it relies on the absence of a label and this can occur if the pod scheduler runs during the delay interval.

One way to address this is to use kubelet's `--register-with-taints` flag. Newly minted nodes can be tainted via the taint `cluster.x-k8s.io=label-sync-pending:NoSchedule`. Assuming workloads don't have this specific toleration, then nothing should be scheduled. KubeadmConfigTemplate provides the means to set taints on nodes (see  JoinConfiguration.NodeRegistration.Taints).

The process of tainting the nodes, can be carried out by the user and can be documented as follows: 

```
If you utilize inequality based selection for workload placement, to prevent unintended scheduling of pods during the initial node startup phase, it is recommend that you specify the following taint in your KubeadmConfigTemplate:

> `cluster.x-k8s.io=label-sync-pending:NoSchedule`

Reviewer 2: ```suggestion
`cluster.x-k8s.io/label-sync-pending:NoSchedule`
```
(same for other occurrences)

```

After the node has come up and the machine controller has applied the labels, the machine controller will also remove this specific taint if it's present. 

In the future, we may consider automating the insertion of the taint via CABPK. 

## Alternatives

### Label propogation from MachineDeployment to Machine

Option #1: Introduce a new field Spec.Node.Labels

Instead of using the top-level metadata, a new field would be introduced:

```
MachineDeployment.Spec.Node.Labels => MachineSet.Spec.Node.Labels => Machine.Spec.Node.Labels
```

The benefit being that it provides a clear indication to the user that these labels will be synced to the Kubernetes Node(s). This will, however, require api changes. 

> Option #2: Propogate labels in the top-level metadata

Reviewer 2: ```suggestion
Option #2: Propagate labels in the top-level metadata
```
nit: please search/replace across the whole doc (also "propagation")
